### PR TITLE
[IMM32] Improve debuggability at all

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Totally improve debuggability.
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Strengthen tracing.
- Introduce `IS_NULL_UNEXPECTEDLY`, `IS_ZERO_UNEXPECTEDLY` etc. macros to manage tracing.
- Add `Imm32UnitTest()` on debug version.

## TODO

- [x] Search and destroy my mistakes.
- [x] Test run.

### CODE REVIEW

**Comment by katahiromz** [comment]
I've triple checked. I will go.

File: dll/win32/imm32/candidate.c [hunk 1]
```diff
@@ -161,27 +161,30 @@ ImmGetCandidateListAW(HIMC hIMC, DWORD dwIndex, LPCANDIDATELIST lpCandList, DWOR
...
+    if (IS_NULL_UNEXPECTEDLY(pCI))
     {
         ImmUnlockIMC(hIMC);
         ImmUnlockClientImc(pClientImc);
         return 0;
     }
 
     if (pCI->dwSize < sizeof(CANDIDATEINFO) || pCI->dwCount <= dwIndex)
+    {
+        ERR("\n");
```
Comment (learn-more) [inline dll/win32/imm32/candidate.c:185]: Why?

Comment (katahiromz) [reply dll/win32/imm32/candidate.c:185]: I think `pCI->dwSize < sizeof(CANDIDATEINFO)` is a logical error.
And `pCI->dwCount <= dwIndex` is a boundary check.

Comment (learn-more) [reply dll/win32/imm32/candidate.c:185]: I meant, why not add anything in the error print message.

Comment (katahiromz) [reply dll/win32/imm32/candidate.c:185]: The program size bloating is a severe problem.
I thought this message was enough for understanding.

Comment (learn-more) [reply dll/win32/imm32/candidate.c:185]: Since when are a few debug messages a severe problem?

Comment (katahiromz) [reply dll/win32/imm32/candidate.c:185]: Do you want to 'normalize' all the comments in `IMM32`?

Comment (katahiromz) [reply dll/win32/imm32/candidate.c:185]: Do you want to 'normalize' all the trace messages in `IMM32`?

Comment (learn-more) [reply dll/win32/imm32/candidate.c:185]: I am interested in the reasoning behind 'debug messages are a problem'

Comment (katahiromz) [reply dll/win32/imm32/candidate.c:185]: https://github.com/reactos/reactos/pull/4807

**Pull request merged**